book/content/material/material_advanced.md

```diff
@@ -0,0 +1,273 @@
+    F.FixedConcentrationBC(right_surf, value=0, species=deuterium),
+    F.FixedConcentrationBC(left_surf, value=1, species=tritium),
+    F.FixedConcentrationBC(right_surf, value=0, species=tritium),
+]
+
+my_model.temperature = 300
+my_model.settings = F.Settings(atol=1e-10, rtol=1e-10, final_time=5)
+my_model.settings.stepsize = F.Stepsize(1)
+my_model.initialise()
+my_model.run()
+
+import matplotlib.pyplot as plt
+
+def plot_profile(species, **kwargs):
+    c = species.post_processing_solution.x.array[:]
+    x = species.post_processing_solution.function_space.mesh.geometry.x[:,0]
+    return plt.plot(x, c, **kwargs)
+
+for species in my_model.species:
+    plot_profile(species, label=species.name)
+
+plt.xlabel('Position')
+plt.ylabel('Concentration')
+plt.legend()
+plt.show()
+```
+
+## Solving a problem with custom material properties ##
+
+Some cases will require user-defined material properties (such as in __[bubbly flows](https://doc.comsol.com/6.0/doc/com.comsol.help.cfd/cfd_ug_fluidflow_multi.09.142.html)__ or __[turbulence-assisted-diffusion](https://en.wikipedia.org/wiki/Turbulent_diffusion)__). 
+
+In this 2D example, we show how to define a spatially-dependent diffusivity. Specifcally, we'll look at a circular diffusivity profile: $ D = 0.1 + x^2 + y^2 $
+
+First, we need to define a 2D mesh:
+
+```{code-cell} ipython3
+import dolfinx
+from dolfinx.mesh import create_unit_square
+from mpi4py import MPI
+import numpy as np
+import festim as F
+from basix.ufl import element
+from dolfinx import plot
+import pyvista
+
+mesh_fenics = create_unit_square(MPI.COMM_WORLD, 20, 20)
+my_model = F.HydrogenTransportProblem()
+my_model.mesh = F.Mesh(mesh_fenics)
+```
+
+Users can define a spatially-dependent material property using `fem.Function`, which requires a function space (with the correct order) to be setup using `fem.functionspace`:
+
+```{code-cell} ipython3
+el = element("Lagrange", mesh_fenics.topology.cell_name(), 2)
+V = dolfinx.fem.functionspace(my_model.mesh.mesh, el)
+
+# Define diffusion coefficient as scalar field
+diffusivity = dolfinx.fem.Function(V)
+diffusivity.interpolate(lambda x: 0.1 + x[0]**2 + x[1]**2)
+```
+
+We can visualize the diffusivity and contours (to verify the circular profile):
+
+```{code-cell} ipython3
+topology, cell_types, geometry = plot.vtk_mesh(V)
+function_grid = pyvista.UnstructuredGrid(topology, cell_types, geometry)
+function_grid.point_data["D"] = diffusivity.x.array.real
+function_grid.set_active_scalars("D")
+
+# Generate contours (e.g. 10 levels)
+contours = function_grid.contour(isosurfaces=5, scalars="D")
+
+plotter = pyvista.Plotter()
+plotter.add_mesh(function_grid, cmap="viridis", show_edges=False, opacity=1)
+plotter.add_mesh(contours, color="r", line_width=0.5)  # contour lines
+plotter.view_xy()
+plotter.add_text("Spatially varying diffusivity", font_size=12)
+
+if not pyvista.OFF_SCREEN:
+    plotter.show()
+```
+
+Now we can add this property directly to a FESTIM `Material`:
```

Comment:
We should add a [note ](https://jupyterbook.org/en/stable/content/content-blocks.html#notes-warnings-and-other-admonitions) here to point out that it's passed to the `D` argument and not `D_0`